linear mixed model

jbloom22 opened the pull request:
Timing on hardcall version of 1kg genomes (2535 samples, 96k variants in kernel, 15.3 million variants in assoc) is 6 minutes, using 344 cores on 1000 partitions (1m on kernel, 4m on assoc).

In this run, I fixed delta at 1000.0 because I'm using a random phenotype that results in fit $delta$ being negligible (outside the search range, where linear regression becomes more appropriate).  But the "fitting delta" step post eigen-decomposition is negligible (~ 1s).

```
gcloud dataproc clusters create cluster-jb --zone us-central1-f --master-machine-type n1-standard-16 --master-boot-disk-size 100 --num-workers 2 --num-preemptible-workers 20 --worker-machine-type n1-standard-16 --worker-boot-disk-size 100 --image-version 1.0 --project broad-ctsa

gcp-hail-lmm1 cluster-jb -l /home/jbloom/hail.log \
  read -i gs://hail-1kg/ALL.1KG.qc.hardcalls.p1000.vds \
  filtervariants expr --keep -c 'va.qc.callRate > .98 && va.qc.AF > 0.005' \
  lmmreg \
    -d 1000.0 \
    -y sa.pheno \
    -c sa.isFemale \
    -k 'v.start % 100 == 0 && va.qc.AF > 0.02 && va.qc.AF < 0.98' \
  showglobals \
    -o gs://jbloom/lmmreg.1kg.globals.json \
  exportvariants \
    -c 'variant = v, pval = va.lmmreg.pval, beta = va.lmmreg.beta, sigmaG2 = va.lmmreg.sigmaG2, chi2 = va.lmmreg.chi2, nNotCalled = va.qc.nNotCalled, nHomRef = va.qc.nHomRef, nHet = va.qc.nHet, nHomVar = va.qc.nHomVar' \
    -o gs://jbloom/lmmreg.1kg.tsv.bgz

merge time: 12.084s
hail: info: timing:
  read: 55.305s
  filtervariants expr: 162.868ms
  lmmreg: 56.056s
  showglobals: 693.315ms
  exportvariants: 4m10.4s
  total: 6m2.6s
```

Timing on 2000 partitions, pre-filtered, 680 cores is 3m42s.

jbloom22 commented:
Made a bunch of tweaks and added docs.

jbloom22 commented:
Docs and a bunch of other updates are done. Ready for review!

tpoterba commented on python/pyhail/dataset.py:
```diff
@@ -689,6 +689,70 @@ def linreg(self, y, covariates="", root="va.linreg", minac=1, minaf=0.0):
 
+    def lmmreg(self, y, kinshipfilter, covariates=None, assocfilter=None, noassoc=None,
+               useml=None, globalroot=None, varoot=None, delta=None,
+               sparsitythreshold=None, forceblock=None, forcegrammian=None):
```
use `False` for boolean args

tpoterba replied:
Use `False` for boolean args, and maybe a list of strings for covariates?

jbloom22 replied:
Changed None to False for Booleans, and empty string for covariates.  I'll refactor all regression methods to list of string in later PR

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala:
```diff
@@ -0,0 +1,167 @@
+    assert(covNames.length == C.cols)
+
+    val useBlock = (options.forceBlock, options.forceGrammian) match {
+      case (false, false) => y.length > 3000 // for small matrices, computeGrammian fits in memory and runs faster than Block
```
should this be parameterized?

jbloom22 replied:
I've chosen a reasonable default of 3000 so users don't have to think about when to use which method.  But I've also exposed the ability to force one method or the other when debugging performance.

danking commented on docs/commands/lmmreg.md:
```diff
@@ -0,0 +1,173 @@
+
+#### Performance
+
+This initial version of `lmmreg` scales efficiently to tens of thousands samples and an unbounded number of variants, making it particularly well-suited to modern sequencing studies. For example, starting from a VDS of the 1000 Genomes Project (consisting of 2535 whole genomes), `lmmreg` computes a kinship matrix based on 100k common variants, fits coefficients and variance components in the samp ...
```
I did a quick calculator for 42 n1-standard-16 workers and 1 hour looks like 34 USD. Where did you get the 50 cents number from?

danking commented on docs/commands/lmmreg.md:
```diff
@@ -0,0 +1,173 @@
+
+Given the eigendecomposition, fitting the global model (Step 4) takes on the order of a few seconds on master. Association testing (Step 5) is fully distributed by variant with per-variant time complexity that is completely independent of the number of sample covariates and dominated by multiplication of the genotype vector $v$ by the matrix of eigenvectors $U^T$ as described below, which we acce ...
+
+Future versions of `lmmreg` will scale to far more samples with exact inference and offer acceleration through approximate inference.
```
I am inclined to leave out future directions from the documentation.

jbloom22 replied:
On the other hand, many users will immediately be asking, when can I use this on my 50k- or 500k-samples dataset.  But I'm also open to taking it out.  I'll get a third opinion.

jbloom22 replied:
Since the paragraph starts with "this initial version of lmmreg", I feel fine taking this out:

"We plan future versions of `lmmreg` will scale to far more samples with exact inference and offer acceleration through approximate inference."

jigold commented on docs/commands/lmmreg.md:
```diff
@@ -0,0 +1,173 @@
+
+and follows a chi-squared distribution with one degree of freedom. Here the ratio $\hat{\sigma}^2_g / \hat{\sigma}_{g,v}^2$ captures the degree to which adding the variant $v$ to the global model reduces the residual phenotypic variance.
+
+#### Kinship: Realized Relationship Matrix (RRM) <a name="RRM"></a>
```
@danking : Correct with the current implementation of the website.

jbloom22 replied:
Ok, I modified, will test it.

danking commented on src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala:
```diff
@@ -0,0 +1,167 @@
+    // FIXME: is this right? and can comma's occur inside backticks?
+    val covNames = options.covSA.trim() match {
+      case "" => List("intercept")
+      case names => "intercept" +: names.split(",").map(_.trim()).toList
```
I think this is correct because all the weird cases should trigger errors in the parser before we get here. It still makes me uncomfortable because there's this implicit dependency that the parser is working the same way this code does.

jbloom22 replied:
for now, I added comment `// covSA has already been successfully parsed`

danking commented on src/main/scala/org/broadinstitute/hail/driver/LinearRegressionCommand.scala:
```diff
@@ -14,7 +14,7 @@ object LinearRegressionCommand extends Command {
 
   class Options extends BaseOptions {
-    @Args4jOption(required = true, name = "-y", aliases = Array("--response"), usage = "Response sample annotation")
+    @Args4jOption(required = true, name = "-y", aliases = Array("--phenotype"), usage = "Response sample annotation")
```
Should the description agree with the option name?

jbloom22 replied:
I changed these back to response.  I think it's better than phenotype, as the covariates are also phenotypes.

danking commented on src/main/scala/org/broadinstitute/hail/stats/ComputeRRM.scala:
```diff
@@ -0,0 +1,82 @@
+object ComputeLocalGrammian {
+  def withoutBlock(A: RowMatrix): DenseMatrix[Double] = {
+    val n = A.numCols().toInt
+    val G = A.computeGramianMatrix().asInstanceOf[SDenseMatrix].values
```
It seems like they hid the `DenseMatrix` type behind `Matrix` on purpose to permit API evolution. I know `Matrix.toArray` isn't overloaded for `DenseMatrix` to do The Right Thing, but do we have performance tests to back up violating API safety?

jbloom22 replied:
I changed them to toArray. These are currently only called once in context of far bigger computation, so avoiding a copy on a local array is not worth violating API safety.

danking commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+              val sb = new SparseGtBuilder()
+              gs.iterator.zipWithIndex.foreach { case (g, i) => if (sampleMaskBc.value(i)) sb.merge(g) }
+              sb.toSparseGtVector(n)
+            }
```
nitpick

I have a minor readability preference for this:

```scala
val SparseGtVectorAndStats(x0, isConstant, maf, nHomRef, nHet, nHomVar, nMissing) = gs.iterator.zipWithIndex
  .filter { case (g, i) => sampleMaskBc.value(i)}
  .foldLeft(new SparseGtBuilder()){ case (b, (g, i)) => b.merge(g) }
  .toSparseGtVector(n)
```

up to you.

cseed commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+    val QtTy = Qt * diagLMM.Ty
+    val TyQtTy = (diagLMM.Ty dot diagLMM.Ty) - (QtTy dot QtTy)
+
+    val assocVariantFilterPred = vds.filterVariantsExprPred(assocFiltExprVA, keep = true)
```
Then you won't need the `filterVariantsExprPred` changes.

jbloom22 replied:
Yes.  Are you recommending I revert those changes too?

cseed replied:
Yes, if you don't need them anymore, don't include the change.

jbloom22 replied:
Done

cseed commented on src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -1108,11 +1108,15 @@ class RichVDS(vds: VariantDataset) {
+  }
+
+  def filterVariantsExpr(cond: String, keep: Boolean): VariantDataset = {
+    val p = filterVariantsExprPred(cond: String, keep: Boolean)
```
You don't need type annotations here.

jbloom22 replied:
Ok

jbloom22 replied:
This is all gone.

cseed commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+      .annotateGlobal(S.data.reverse: IndexedSeq[Double], TArray(TDouble), s"$rootGA.evals")
+
+    val vds2 = if (optDelta.isEmpty) {
+      val maxLogLkhd = diagLMM.maxLogLkhd.get
```
Matching on `diagLMM.maxLogLkhd` is safer.  Avoid `isEmpty` and `get`.

jbloom22 replied:
Will do.

jbloom22 replied:
I changed to this:
```
    val vds2 = (diagLMM.maxLogLkhd, diagLMM.deltaGridLogLkhd) match {
      case (Some(maxLogLkhd), Some(deltaGridLogLkhd)) =>
        val (logDeltaGrid, logLkhdVals) = deltaGridLogLkhd.unzip
        vds1.annotateGlobal(
          Row(maxLogLkhd, logDeltaGrid, logLkhdVals),
          TStruct(("maxLogLkhd", TDouble), ("logDeltaGrid", TArray(TDouble)), ("logLkhdVals", TArray(TDouble))), rootGA)
      case _ =>
        assert(optDelta.isDefined)
        vds1
    }
```

cseed commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+    val TyQtTy = (diagLMM.Ty dot diagLMM.Ty) - (QtTy dot QtTy)
+
+    val assocVariantFilterPred = vds.filterVariantsExprPred(assocFiltExprVA, keep = true)
+    val keepAll = assocFiltExprVA == "true"
```
Why special case this?

jbloom22 replied:
Because in practice many users don't care about LOCO, so they do filter the vds first and don't need any addition assocFilter.  In that case, I was trying to make things a bit more efficient by not bothering to evaluate the condition each time (probably over-optimizing).  This will go away once vds is pre-filtered.

cseed replied:
Yes, over-optimizing.  And this it will have to change as we improve the expr implementation.

jbloom22 replied:
Gone

cseed commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+      .annotateGlobal(globalSe2: Double, TDouble, s"$rootGA.sigmaE2")
+      .annotateGlobal(delta: Double, TDouble, s"$rootGA.delta")
+      .annotateGlobal(h2: Double, TDouble, s"$rootGA.h2")
+      .annotateGlobal(S.data.reverse: IndexedSeq[Double], TArray(TDouble), s"$rootGA.evals")
```
You could also write these as a single insert of a struct.

jbloom22 replied:
Will do.

jbloom22 replied:
Done

cseed commented on python/pyhail/dataset.py:
```diff
@@ -689,7 +689,71 @@ def linreg(self, y, covariates="", root="va.linreg", minac=1, minaf=0.0):
-    def logreg(self, test, y, covariates=None, root=None):
+    def lmmreg(self, y, kinshipfilter, covariates="", assocfilter=None, noassoc=False,
+               useml=False, globalroot=None, varoot=None, delta=None,
+               sparsitythreshold=1.0, forceblock=False, forcegrammian=False):
```
If the lmm command takes the kinship vds by name, the kinship `VariantDataset` should be given as an argument like `join`.

jbloom22 replied:
ok

jbloom22 replied:
Done

cseed commented on src/main/scala/org/broadinstitute/hail/stats/ComputeRRM.scala:
```diff
@@ -0,0 +1,82 @@
+    (ComputeLocalGrammian.withoutBlock(A) :* mRec, A.numRows().toInt)
+  }
+
+  def withBlock(vds: VariantDataset): (DenseMatrix[Double], Int) = {
```
More concise to just have a `useBlock: Boolean` parameter than duplicating these functions.

jbloom22 replied:
Will do.

jbloom22 replied:
Done

cseed commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+
+    val n = y.length
+    val c = C.cols
+    val shift =
```
Why not put shift inside the ML and REML functions?  Then you won't need the conditional.

jbloom22 replied:
Because the shift should only be evaluated once.  It doesn't depend on delta.

cseed replied:
Then only compute it once. That's what closures are for.

jbloom22 replied:
Gotcha

cseed commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+      else
+        n + (n - c) * math.log(2 * math.Pi) - logdet(C.t * C)._2
+
+    object LogLkhdML extends UnivariateFunction {
```
Again, what's with the objects?  Are you using `UnivariateFunction`?  Why not just define `def logLkhdML(delta: Double): Double = ...`?

jbloom22 replied:
In order to use the Brent optimizer

cseed commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+
+    val searchInterval = new SearchInterval(-10d, 10d, approxLogDelta)
+    val goal = GoalType.MAXIMIZE
+    val objectiveFunction = new UnivariateObjectiveFunction(if (useML) LogLkhdML else LogLkhdREML)
```
Ah, I see now.  It's fine as it is now, or you could do:

```
  new UnivariableObjectFunction(new UnivariantFunction {
    def value(x: Double): Double = if (useML) logLkhdML(x) else logLkhdREML(x))
  }
```

jbloom22 replied:
I like your pattern more, will change it.

danking replied:
Not sure if it matters but pushing the `if` into the `UnivariantFunction.apply` means it gets checked for each call to `value`, which seems not ideal for the objective function.

jbloom22 replied:
ah yes! That's actually why I did it this way.

cseed replied:
In fact, I'd just make an implicit conversion from `(Double) => Double` to `UnivariantFunction`.  That's all it is, anyway.

cseed replied:
And the implicit conversion will handle the if efficiently.

cseed commented on src/main/scala/org/broadinstitute/hail/stats/SparseGtBuilder.scala:
```diff
@@ -0,0 +1,66 @@
+
+// constructs SparseVector of genotype calls with missing values mean-imputed
+// if all genotypes are missing then all elements are NaN
+class SparseGtBuilder extends Serializable {
```
This looks nearly identical to `LinRegBuilder`.  Is it really necessarily to duplicate all this?

jbloom22 replied:
Good call. It didn't start off this similar, but evolved in that direction so I'll take another look.

jbloom22 replied:
SparseGtBuilder is more generic and only depends on genotypes, whereas LinRegBuilder is optimized to precompute sufficient stats like xx and xy needed only in case of straight up linear regression.  I think I should return to this when unifying what the regression models output (right now LMM gives additional counts like nHet with respect to included samples only, but if may make more sense to filterSamples on the vds at the beginning, will discuss more in person).

jbloom22 commented on src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala:
```diff
@@ -15,13 +15,13 @@ object LogisticRegressionCommand extends Command {
 
   class Options extends BaseOptions {
-    @Args4jOption(required = true, name = "-t", aliases = Array("--test"), metaVar = "TEST",  usage = "Statistical test: `wald', `lrt', or `score'")
+    @Args4jOption(required = true, name = "-t", aliases = Array("--test"), usage = "Statistical test: `wald', `lrt', or `score'")
```
At first I started adding more metaVars, and then I I decided I wanted to be consistent on usage across the the three regression models and wasn't yet sure of plan so decided to rip out all metaVars for now and then come back in another PR across all three.

cseed replied:
OK, reasonable.  This is all going away soon anyway.

jbloom22 commented on src/main/scala/org/broadinstitute/hail/stats/ComputeRRM.scala:
```diff
@@ -0,0 +1,82 @@
+
+// mean 0, norm sqrt(n), variance 1 (constant variants are None)
+object ToNormalizedGtArray {
+  def apply(gs: Iterable[Genotype], nSamples: Int): Option[Array[Double]] = {
```
Can you elaborate? What should the type of the function in Stats be?

cseed replied:
`def toNormalizedGtArray(gs: Iterable[Genotype], nSamples): Option[Array[Double]] = ...`.

jbloom22 replied:
Done

jbloom22 commented on src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala:
```diff
@@ -0,0 +1,167 @@
+
+    val ec = EvalContext(symTab)
+
+    def toDouble(t: BaseType, code: String): Any => Double = t match {
```
The Boolean to Double conversion is defined in `RichBoolean` in  `org.broadinstitute.hail.utils.richUtils`.  That said, in a latter pass, I intend to refactor the extraction of covariates to a common class for all regressions implemented so far.

jbloom22 commented on src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala:
```diff
@@ -0,0 +1,167 @@
+    val vds = state.vds
+
+    val pathVA = Parser.parseAnnotationRoot(options.rootVA, Annotation.VARIANT_HEAD)
+    val pathGA = Parser.parseAnnotationRoot(options.rootGA, Annotation.GLOBAL_HEAD) // FIXME: included only to check for errors
```
Done

jbloom22 commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+    noAssoc: Boolean = false,
+    useML: Boolean = false,
+    sparsityThreshold: Double = 1d,
+    useBlock: Boolean = true): VariantDataset = {
```
Agreed, I removed all defaults from LinearMixedModel apply

jbloom22 commented on src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala:
```diff
@@ -0,0 +1,167 @@
+    if (d < 1)
+      fatal(s"$n samples and $k ${plural(k, "covariate")} with intercept implies $d degrees of freedom.")
+
+    info(s"lmmreg: running lmmreg on $n samples with $k sample ${plural(k, "covariate")}...")
```
I moved it inside LinearMixedModel

jbloom22 commented on src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala:
```diff
@@ -0,0 +1,167 @@
+    var covSA: String = ""
+
+    @Args4jOption(required = true, name = "-k", aliases = Array("--kinshipfilter"), usage = "Variant filter expression for the kinship matrix")
+    var kinshipFiltExprVA: String = _
```
done

jbloom22 commented on src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala:
```diff
@@ -0,0 +1,167 @@
+    var kinshipFiltExprVA: String = _
+
+    @Args4jOption(name = "-a", aliases = Array("--assocfilter"), usage = "Variant filter expression for association testing")
+    var assocFiltExprVA: String = _
```
done

jbloom22 commented on src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala:
```diff
@@ -0,0 +1,303 @@
+    if (approxLogDelta == logmin)
+      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta lower search boundary e^$logmin = ${math.exp(logmin)}, indicating negligible enviromental component of variance. The model is likely ill-specified.")
+    else if (approxLogDelta == logmax)
+      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta upper search boundary e^$logmax = ${math.exp(logmax)}, indicating negligible genetic component of variance. Standard linear regression may be more appropriate.")
```
Let's chat about this.

danking commented:
Modulo the exception/`Either` suggestion in `LinearMixedModel.scala`, I'm happy with this. We could push that refactoring into a follow-up pull request.

jbloom22 commented:
@cseed I've moved LMM doc to Python API.  The output is still a vds rather than a KeyTable, but this can be merged in the meantime if it passes your review as well.

jbloom22 commented:
I've rebased `jbloom22:lmm_getthisin` onto `jbloom22:py_reg` (which should go in first) and made the parallel modifications (LinearMixedModelCommand is gone, command line relics gone, refactored using RegressionUtils, tests modified to accommodate changing interface, etc).  I'll do some command line testing next to be safe.

jbloom22 commented:
example python code

```
import pyhail
hc = pyhail.HailContext()
vds = (hc.import_vcf('src/test/resources/sample.vcf')
    .split_multi()
    .variant_qc()
    .annotate_samples_expr('sa.culprit = gs.filter(g => v == Variant("20", 13753124, "A", "C")).map(g => g.gt).collect()[0]')
    .annotate_samples_expr('sa.pheno = rnorm(1,1) * sa.culprit')
    .annotate_samples_expr('sa.cov1 = rnorm(0,1)')
    .annotate_samples_expr('sa.cov2 = rnorm(0,1)')
    .linreg('sa.pheno', ['sa.cov1', 'sa.cov2']))

vds_kinship = vds.filter_variants_expr('va.qc.AF > .05')

vds = vds.lmmreg(vds_kinship, 'sa.pheno', ['sa.cov1', 'sa.cov2'])

vds.export_variants('lmmreg.tsv', 'Variant = v, lin_beta = va.linreg.beta, lmm_beta = va.lmmreg.beta, lin_pval = va.linreg.pval, lmm_pval = va.lmmreg.pval, va.lmmreg.*')
```

jbloom22 commented:
I've rebased to move from pyhail to hail.

jbloom22 commented:
Rebased! Going to run a test today to see if results change on DataProc with and without natives.

jbloom22 commented:
Rebased again, tiny improvements
